src/main/java/org/truffleruby/core/string/StringSupport.java

```diff
@@ -1592,22 +1638,26 @@ public static boolean upcaseMultiByteComplex(Encoding enc, CodeRange originalCod
      * the string doesn't require changes. The encoding must be ASCII-compatible (i.e. represent each ASCII character as
      * a single byte ({@link Encoding#isAsciiCompatible()}). */
     @TruffleBoundary
-    public static byte[] capitalizeMultiByteAsciiSimple(Encoding enc, CodeRange codeRange, byte[] bytes) {
+    public static byte[] capitalizeMultiByteAsciiSimple(Encoding enc, TruffleString.CodeRange codeRange,
+            InternalByteArray byteArray) {
         assert enc.isAsciiCompatible();
         boolean modified = false;
-        final int end = bytes.length;
 
-        if (end == 0) {
+        int p = byteArray.getOffset();
+        final int end = byteArray.getEnd();
+        var bytes = byteArray.getArray();
+
+        if (byteArray.getLength() == 0) {
             return bytes;
         }
 
-        if (StringSupport.isAsciiLowercase(bytes[0])) {
+        if (StringSupport.isAsciiLowercase(bytes[p])) {
             bytes = bytes.clone();
```

Comment:
It also seems best to only copy the relevant "slice" here (between offset and end) with CopyToByteArrayNode (same below).
The caller code seems to correctly handle the offset in this case though.

Reply:
Fixed in 2c694299877d8d45de6a196e95c2a50009ccacfe.